infer/default_provider.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2016-2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infer
+
+import (
+	"fmt"
+
+	provider "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/middleware/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+)
+
+type DefaultProvider struct {
+	name, version string
+	metadata      schema.Metadata
+	resources     []InferredResource
+	components    []InferredComponent
+	functions     []InferredFunction
+	config        InferredConfig
+	moduleMap     map[tokens.ModuleName]tokens.ModuleName
+}
+
+// NewDefaultProvider creates an inferred provider which fills as many defaults as possible.
+//
+// A base set of defaults are provided to create a minimal provider, but can be initially overridden by
+// passing in an [Options] struct. Further customization can be done by chaining method calls
+// on the returned [DefaultProvider] object.
+//
+// This is an example of how to create a simple provider with a single component resource:
+//
+//	type RandomComponent struct {
+//		pulumi.ResourceState
+//		RandomComponentArgs
+//	 	Password        pulumi.StringOutput `pulumi:"password"`
+//	}
+//
+//	type RandomComponentArgs struct {
+//		Length pulumi.IntInput `pulumi:"length"`
+//	}
+//
+//	func NewMyComponent(ctx *pulumi.Context, name string,
+//			compArgs RandomComponentArgs, opts ...pulumi.ResourceOption) (*RandomComponent, error) {
+//		// Define your component constructor logic here.
+//	}
+//
+//	func main() {
+//		err := infer.NewDefaultProvider(nil).
+//			WithName("go-components").
+//			WithVersion("v0.0.1").
+//			WithComponents(
+//				infer.Component(NewMyComponent),
+//			).
+//			BuildAndRun()
+//	}
+func NewDefaultProvider(opts *Options) *DefaultProvider {
```

Comment:
We should probably mention that things like the language map may change between versions of pulumi-go-provider.